### PR TITLE
Adapter enhancements in Astro 4.2

### PR DESCRIPTION
**Pull request by lilnasy**
#### Description (required)
Documents new options in the adapter API being introduced in Astro 4.2

#### Related issues & labels (optional)
- https://github.com/withastro/astro/pull/9661

### CODE REVIEW

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 1]
```diff
@@ -194,18 +194,54 @@ export function start(manifest) {
...
+const response = await app.render(request, { addCookieHeader: true });
+```
+
+###### `clientAddress`
+
+The client IP address that will be made available as `Astro.clientAddress` in pages, and as `ctx.clientAddress` in API routes and middleware.
+
+The example below reads `x-forwarded-for` header and passes it as `clientAddress`, Astro makes this value available to the user as `Astro.clientAddress`.
+```js
+const clientAddress = request.headers.get("x-private-header");
```
Comment (alexnguyennz) [inline src/content/docs/en/reference/adapter-reference.mdx:225]:
```suggestion
const clientAddress = request.headers.get("x-forwarded-for");
```

Comment (alexnguyennz) [reply src/content/docs/en/reference/adapter-reference.mdx:225]: From `The example below reads `x-forwarded-for` header` (or the other way around)

Comment (lilnasy) [reply src/content/docs/en/reference/adapter-reference.mdx:225]: That's exactly right, thanks!

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 2]
```diff
@@ -194,18 +194,54 @@ export function start(manifest) {
...
+
+###### `clientAddress`
+
+The client IP address that will be made available as `Astro.clientAddress` in pages, and as `ctx.clientAddress` in API routes and middleware.
+
+The example below reads `x-forwarded-for` header and passes it as `clientAddress`, Astro makes this value available to the user as `Astro.clientAddress`.
+```js
+const clientAddress = request.headers.get("x-forwarded-for");
+const response = await app.render(request, { clientAddress });
+```
```
Comment (VoxelMC) [inline src/content/docs/en/reference/adapter-reference.mdx:227]:
``````suggestion
```js {clientAddress}
const clientAddress = request.headers.get("x-forwarded-for");
const response = await app.render(request, { clientAddress });
```
``````

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 3]
```diff
@@ -194,18 +194,54 @@ export function start(manifest) {
...
 
+```js
+const routeData = app.match(request);
+if (routeData) {
+    return app.render(request, { routeData });
+} else {
+    /* adapter-specific 404 response */
+    return new Response(..., { status: 404 });
+}
+```
```
Comment (VoxelMC) [inline src/content/docs/en/reference/adapter-reference.mdx:241]:
``````suggestion
```js {routeData}
const routeData = app.match(request);
if (routeData) {
    return app.render(request, { routeData });
} else {
    /* adapter-specific 404 response */
    return new Response(..., { status: 404 });
}
```
``````

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 4]
```diff
@@ -194,18 +194,54 @@ export function start(manifest) {
...
+
+###### `locals`
+
 The example below reads a header named `x-private-header`, attempts to parse it as an object and pass it to `locals`, which can then be passed to any [middleware function](/en/guides/middleware/).
 
 ```js
```
Comment (VoxelMC) [inline src/content/docs/en/reference/adapter-reference.mdx:247]:
``````suggestion
```js {locals}
``````

**Review by VoxelMC — COMMENTED** [review]
Hello!!

I know it's still a draft, but should we entertain the idea of having some Expressive Code here? Since we are not inlining the options in the function, it might be helpful if the eyes have somewhere to snap to!

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 5]
```diff
@@ -194,18 +194,54 @@ export function start(manifest) {
...
 ```
 
-The method accepts a mandatory `request` argument, and an optional argument for [`routeData`](/en/reference/integrations-reference/#routedata-type-reference) and [`locals`](/en/reference/api-reference/#astrolocals).
+##### `RenderOptions`
+
+The `app.render()` method accepts a mandatory `request` argument, and an optional argument for [`clientAddress`](/en/reference/api-reference/#astroclientaddress),[`locals`](/en/reference/api-reference/#astrolocals), [`routeData`](/en/reference/integrations-reference/#routedata-type-reference).
+
+###### `addCookieHeader`
+
+Whether to automatically add all cookies written by `Astro.cookie.set()` to the response headers.
```
Comment (sarah11918) [inline src/content/docs/en/reference/adapter-reference.mdx:211]:
```suggestion
Whether or not to automatically add all cookies written by `Astro.cookie.set()` to the response headers.

```

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 6]
```diff
@@ -194,18 +194,54 @@ export function start(manifest) {
...
+When set to `true`, they will be added to the `Set-Cookie` header of the response as comma-separated key=value pairs. You can use the standard `response.headers.getSetCookie()` API to read them individually.
```
Comment (sarah11918) [inline src/content/docs/en/reference/adapter-reference.mdx:212]:
```suggestion
When set to `true`, they will be added to the `Set-Cookie` header of the response as comma separated key-value pairs. You can use the standard `response.headers.getSetCookie()` API to read them individually.
```
I'm going to need a @Yan-Thomas check in aisle 5!

- comma separated: no hypen?
- key-value: hypen?
- "comma separated key-value pairs" no comma?

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 7]
```diff
@@ -194,18 +194,54 @@ export function start(manifest) {
...
+
+```js
+const response = await app.render(request, { addCookieHeader: true });
+```
+
+###### `clientAddress`
+
+The client IP address that will be made available as `Astro.clientAddress` in pages, and as `ctx.clientAddress` in API routes and middleware.
+
+The example below reads `x-forwarded-for` header and passes it as `clientAddress`, Astro makes this value available to the user as `Astro.clientAddress`.
```
Comment (sarah11918) [inline src/content/docs/en/reference/adapter-reference.mdx:223]:
```suggestion
The example below reads the `x-forwarded-for` header and passes it as `clientAddress`. This value is available to the user as `Astro.clientAddress`.
```
What is more correct:

> The example below reads THE `x-forewarded-for` header
> The example below reads `x-forwarded-for` FROM/IN THE header
... something else?

File: src/content/docs/en/reference/adapter-reference.mdx (hunk 2 above)
Comment (sarah11918) [reply src/content/docs/en/reference/adapter-reference.mdx:227]: I agree with the sentiment and approve, but the syntax is quotation marks, not braces, for text. Braces are used for line numbers!

File: src/content/docs/en/reference/adapter-reference.mdx (hunk 3 above)
Comment (sarah11918) [reply src/content/docs/en/reference/adapter-reference.mdx:241]: Same comment as above! I like the highlighting addition! needs to be `"routeData"` syntax!

File: src/content/docs/en/reference/adapter-reference.mdx (hunk 4 above)
Comment (sarah11918) [reply src/content/docs/en/reference/adapter-reference.mdx:247]: Would also approve here, with the right syntax!

**Review by sarah11918 — COMMENTED** [review]
Just some minor polishing here, @lilnasy ! Not much work at all!

Just to double-triple check: did you also go through closely all the code examples on this page and update any that might be affected? (I don't see any updates/corrects in code, so does that mean we were good?)

File: src/content/docs/en/reference/adapter-reference.mdx (hunk 7 above)
Comment (lilnasy) [reply src/content/docs/en/reference/adapter-reference.mdx:223]: The first one, I'm pretty sure.

**Comment by lilnasy** [comment]
> did you also go through closely all the code examples on this page
Not yet. I meant to take a look at the deployed preview for a final look. I'll do that now.

File: src/content/docs/en/reference/adapter-reference.mdx (hunk 2 above)
Comment (VoxelMC) [reply src/content/docs/en/reference/adapter-reference.mdx]: Oops, brain fart!

**Comment by lilnasy** [comment]
There seems to be nothing else to update on the page. I do want to order `RenderOptions` alphabetically.

File: src/content/docs/en/reference/adapter-reference.mdx (hunk 6 above)
Comment (VoxelMC) [reply src/content/docs/en/reference/adapter-reference.mdx:213]: > I'm going to need a @Yan-Thomas check in aisle 5!
> 
> * comma separated: no hypen?
> * key-value: hypen?
> * "comma separated key-value pairs" no comma?

If I may, I believe it should be: 
``they will be added to the `Set-Cookie` header of the response as comma-separated key-value pairs.``

Comment (lilnasy) [reply src/content/docs/en/reference/adapter-reference.mdx:213]: We could just be less specific. We are not manually performing this step, a standard API does it for us.

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 8]
```diff
@@ -194,21 +194,45 @@ export function start(manifest) {
...
+
+The example below reads the `x-forwarded-for` header and passes it as `clientAddress`. This value becomes available to the user as `Astro.clientAddress`.
+
+```js "clientAddress"
+const clientAddress = request.headers.get("x-forwarded-for");
+const response = await app.render(request, { clientAddress });
+```
+
+###### `locals`
+
```
Comment (sarah11918) [inline src/content/docs/en/reference/adapter-reference.mdx:232]:
```suggestion

The [`context.locals` object](/en/reference/api-reference/#contextlocals) used to store and access information during the lifecycle of a request.
```

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 9]
```diff
@@ -194,21 +194,46 @@ export function start(manifest) {
...
+The [`context.locals` object](/en/reference/api-reference/#contextlocals) used to store and access information during the lifecycle of a request.
```
Comment (sarah11918) [inline src/content/docs/en/reference/adapter-reference.mdx:233]:
```suggestion
The [`context.locals` object](/en/reference/api-reference/#contextlocals) used to store and access information during the lifecycle of a request.

```

**Review by sarah11918 — APPROVED** [review]
@lilnasy Would love one final review from you on this, as I made some tiny adjustments (e.g. linked within-page to headings, and then linked *those* to other pages)

But on my end, this feels good to go!

**Comment by lilnasy** [comment]
I was watching 👀. Love the edit around `RenderOptions`. Looks great to me!

**Review by VoxelMC — COMMENTED** [review]
Great work so far!

Just some cleanup here, let me know if it's incorrect at all!

File: src/content/docs/en/reference/adapter-reference.mdx [hunk 10]
```diff
@@ -219,7 +219,7 @@ const response = await app.render(request, { addCookieHeader: true });
 
 ###### `clientAddress`
 
-The client IP address that will be made available as `Astro.clientAddress` in pages, and as `ctx.clientAddress` in API routes and middleware.
+The client IP address that will be made available as []`Astro.clientAddress`](/en/reference/api-reference/#astroclientaddress) in pages, and as `ctx.clientAddress` in API routes and middleware.
```
[the remark this answers is not in the document]
Comment (lilnasy) [reply src/content/docs/en/reference/adapter-reference.mdx]: This was fixed immediately!

**Pull request merged**